### PR TITLE
added an error message if parameters are not loaded

### PR DESCRIPTION
**Pull request by superjax**
This addresses #342 

Needs to be Tested
 - [x] hardware
 - [x] SIL

### CODE REVIEW

**Comment by dpkoch** [comment]
Cool. I'm wondering though if changing the message to something like "Reverting to default parameters" would be more informative?

**Comment by plusk01** [comment]
Actually, it looks like this doesn't work in its current form because of initialization order:

```c++
void ROSflight::init()
{
  // Initialize the arming finite state machine
  state_manager_.init();

  // Read EEPROM to get initial params
  params_.init();

  // Initialize Mixer
  mixer_.init();

  /***********************/
  /***  Hardware Setup ***/
  /***********************/

  // Initialize PWM and RC
  rc_.init();

  // Initialize MAVlink Communication
  comm_manager_.init();

  // ...
}
```

Is it worth adding some sort of latched warning/error mechanism?

**Comment by superjax** [comment]
Hmm.  Darn, I was hoping to avoid that.

Okay, well what module do you think would be responsible for that?  Maybe the CommLink?  He could have a buffer of messages to send when he gets a connection

That would mean that the CommLink would have to be initialized first.  I can't remember if that is possible or not.  However, based on this discussion, I think it needs to be first, anyway.

**Comment by dpkoch** [comment]
One possible method: The comm manager could keep a small circular buffer of log messages to send. Right now in the log() function, the message is immediately sent. We could change it so that if the comm manager has not yet been initialized and received a heartbeat (which indicates a connection has been made) it caches the message, otherwise it sends it immediately. Then in the heartbeat callback, the first time it's called it sends any cached log messages.

This is very similar to how the hardfault handling is currently implemented. It's a bit of work for a fairly small increase in functionality, but it probably improves the user experience pretty significantly so may be worth it. What do you guys think?

**Comment by plusk01** [comment]
Yeah, I was thinking about a buffer, so I like that idea. I can take care of that.

How many messages are we willing to buffer? 3 or 4? (x 50 bytes).

**Comment by plusk01** [comment]
the buffer seems to work, but it is being filled with messages from `rc_.init()`:

```bash
[ INFO] [1561839984.939654203]: Got HEARTBEAT, connected.
[ERROR] [1561839984.939723851]: Autopilot ERROR: Invalid mixer
[ERROR] [1561839984.939766135]: Autopilot ERROR: Uncalibrated IMU
[ WARN] [1561839984.939798835]: Autopilot now in ANGLE mode
[ INFO] [1561839984.945727113]: Firmware version: v1.1.0-302-g39909ef9-dirty
[ WARN] [1561839984.945767485]: [Autopilot]: Unable to load parameters; using default values
[ERROR] [1561839984.945792273]: [Autopilot]: Invalid Mixer Choice
[ERROR] [1561839984.945812382]: [Autopilot]: Invalid Mixer Choice
[ INFO] [1561839984.945832998]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.945852901]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.945871821]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.945890041]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.945911454]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.945933887]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.945954122]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.945982013]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946007121]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946034664]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946061603]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946082490]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946115968]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946146040]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946172349]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946191956]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946211599]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946237081]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946263330]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946283270]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946308967]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946329878]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946348287]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946366364]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946392342]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946420082]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946439581]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946458171]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946476209]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946495276]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946513882]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946533325]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946557489]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946579345]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946598547]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946616877]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946635557]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946654358]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946672604]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946690602]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946708987]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946728389]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946748759]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946773379]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946793325]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946814139]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946839241]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946865820]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946885548]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946905901]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946925119]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.946945743]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.946974421]: [Autopilot]: ARM switch not mapped
[ INFO] [1561839984.946996219]: [Autopilot]: ATTITUDE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.947022290]: [Autopilot]: THROTTLE OVERRIDE switch mapped to RC channel 4
[ INFO] [1561839984.947044430]: [Autopilot]: ATTITUDE TYPE switch not mapped
[ INFO] [1561839984.947064420]: [Autopilot]: Booting
```

This would require a buffer of ~60 messages (300 bytes) to see the params warning log.

Maybe a better question is are the RC messages necessary / can they be sanely throttled?

**Comment by superjax** [comment]
I think that throttling these messages makes a lot of sense, and I'm wondering if this is related to our getting the `ARM_CHANNEL` parameter every time we change any parameter from `rosflight_io`.  I wonder if there is some sort of weird loop in the RC parameter settings.

**Comment by dpkoch** [comment]
It appears that these messages are getting sent so often because all of the parameter callbacks for the RC module are getting called at startup, which then call `init_rc()` a bunch of times and sends those messages. We could probably clean that up to solve this problem

**Comment by dpkoch** [comment]
The issue with the RC module re-initializing for every parameter callback it subscribes to (and spamming the messages as a result) should be fixed by #337

**Comment by plusk01** [comment]
cool, I'll take a look at it

File: src/comm_manager.cpp [hunk 1]
```diff
@@ -350,18 +381,26 @@ void CommManager::attitude_correction_callback(const turbomath::Quaternion &q)
...
+    if (this->RF_.board_.has_backup_data())
+    {
+        this->send_error_data();
+    }
   }
-  this->send_heartbeat();//respond to heartbeats with a heartbeat
+
+  /// JSJ: I don't think we need this
+  // respond to heartbeats with a heartbeat
+  this->send_heartbeat();
```
Comment (plusk01) [inline src/comm_manager.cpp]: This *is* needed, right?

@dpkoch

**Pull request merged**